Eigenlayer | Ethereum (#188)
deficollective/defiscan: src/content/protocols/eigenlayer/ethereum.md

# Ratings

## Chain

The report is concerned with the Eigenlayer protocol deployed on Ethereum mainnet. Ethereum mainnet achieves a _Low_ centralization risk score.

> Chain score: Low

## Upgradeability

Contracts which control user funds, the accounting of delegation and slashing are fully _upgradeable_ and this can lead to _loss of user funds_ and _loss of unclaimed yield_.

The withdrawal process can be paused immediately by pausers registered in the `PauserRegistry` contract. Pausing can lead to temporary loss of access to funds, the [Community Council (9/13)](#security-council) however can resume the pause immediately if misused.

> Rewards committed by _AVSs_ to the `RewardsCoordinator` contract are distributed to _Operators_ and _Stakers_ via a merkle tree root posted by the `RewardsUpdater` role owner, which computes eligibility for each _(Re)Staker_ and _Operator_ based on their stake and _AVS_ rewards specification. If the account that holds the `RewardsUpdater` role is compromised, the rewards can be distributed to the wrong accounts, which leads to _loss of unclaimed yield_.

Reviewer: I would rephrase as such:

Repeating the role with owner vs holder overcomplicates a little bit.

```suggestion
Rewards committed by _AVSs_ to the `RewardsCoordinator` contract are distributed to _Operators_ and _Stakers_ via a merkle tree root posted by an account with the role `RewardsUpdater` , which computes eligibility for each _(Re)Staker_ and _Operator_ based on their stake and _AVS_ rewards specification. If an account with that role is compromised, the rewards can be distributed to the wrong addresses, which leads to _loss of unclaimed yield_.
```

> Upgradeability score: High

## Autonomy

The Eigenlayer protocol does not rely on a single _Operator_, nor on a single _AVS_ to work since it is a market place that brings _AVSs_ and _Operators_ together. Thus the overall _Dependency_ risk is _low_. However, single _AVSs_ or _Operators_ can impose risks to each other as the respective counterparties.

The [Dependencies](#dependencies) section goes into more detail about risks potentially originating from _AVSs_ and _Operators_.

> Autonomy score: Low

## Exit Window

### Upgrading Contracts

> Upgrading of all upgradeable Eigenlayer smart contracts can be executed through the [Executor Multisig (1/2)](#security-council) which has two signers, the [Community Multisig (9/13)](#security-council) and the [Protocol TimelockController](#security-council). The [Protocol TimelockController](#security-council) has an _Exit Window_ of 10 days. The upgrade can also be executed immediately through the [Community Council (9/13)](#security-council) signer which adheres to the [Security Council Requirements](/learn-more#security-council-requirements).

Reviewer: Are there different delays between the executive multisig and the community one? Why mention the latter twice?

Author: yes they have different delayes (none vs 10 days and none vs 24 days)
[7919123](https://github.com/deficollective/defiscan/pull/188/commits/7919123368d836780b8d3209e2a0bd71d3a0eb44)

> For the `bEIGEN` token there is a dedicated parallel upgrading process, which is also controlled by the [Community Council (9/13)](#security-council) signer and a dedicated [bEIGEN TimelockController](#security-council) with an _Exit Window_ of 24 days.

Reviewer: You haven't mentioned the `bEIGEN` token before this point, it would be good to say what it is and why it's used

### Pausing Strategy Contracts

> Single strategies can be paused immediately by pausers registered in the `PauserRegistry` contract which includes the [Pauser Multisig (1/6)](#security-council), the [Executor Multisig (1/2)](#security-council) and the [Operations Multisig (3/6)](#security-council). During a pause, the _(Re)Stakers_ cannot withdraw their funds. Resuming can only be enforced by the [Executor Multisig (1/2)](#security-council) with a 10 day _Exit Window_ or immediately through the [Community Council (9/13)](#security-council) signer which adheres to the [Security Council Requirements](/learn-more#security-council-requirements).

Reviewer: Same here, what exactly are strategies?

### Rewards

> The `RewardsUpdater` can immediately remove a root and post a new root, claiming all deposited funds by _AVSs_ instead of distributing them to _Operators_ and _Stakers_. Immediate _loss of unclaimed yield_ leads to _Medium_ _Exit Window_ score.

Reviewer: Could you reiterate that the root holds the information of rewards attribution

### Overall Exit Window

The worst _Exit Window_ score of the different vectors is _High_. Immediate upgrade of contracts can only be executed by the [Community Council (9/13)](#security-council) signer, thus the _Exit Window_ score is _High_. However, the impact on the total score is _Medium_ (_Stage 1_), as the [Community Council (9/13)](#security-council) satisfies the [Security Council Requirements](/learn-more#security-council-requirements). Multisigs that are not compliant with the [Security Council Requirements](/learn-more#security-council-requirements) can initiate upgrades as well, but are protected by an _Exit Window_ of more than 7 days and can be overruled by the [Community Council (9/13)](#security-council).

> > Exit Window score: High

Reviewer: So shouldn't the score should be Medium then?

Author: discuss with Nils

# Protocol Analysis

![Eigenlayer](../diagrams/eigenlayer-overview.png)

## Roles

### Stakers and Restakers

A _Staker_ or _Restaker_ is any user who has assets deposited (or "restaked") into EigenLayer. They can stake or restake multiple assets that are accepted as _Economic Security_ to secure _AVSs_.

The _Stakers_ and _Restakers_ have the following actions available; 1) deposit assets into EigenLayer via either the `StrategyManager` (for ERC20s) or the `EigenPodManager` (for Beacon Chain ETH), 2) withdraw assets via the `DelegationManager`, no matter what assets they're withdrawing, 3) delegate to an _Operator_ via the `DelegationManager`, 4) undelegating from an _Operator_ via the `DelegationManager`, 5) and claiming rewards posted by _AVSs_ from the `RewardsCoordinator` contract.

### Operators

_Operators_ are entities that opt-in to _Operator Sets_ created by _AVSs_ to execute tasks for the _AVS_. _Operators_ receive economic stake delegated by the _Stakers_ and _Restakers_ and take a cut from the weekly rewards issued by the _AVSs_ (default: 10% _Operator_, 90% _Stakers_ and _Restakers_). _Operators_ can also stake themselves, the roles are not mutually exclusive.

> _Operators_ have the following actions available; 1) register to an _Operator Set_ via the `AllocationManager` contract, 2) deregister from an _Operator Set_ via the `AllocationManager` contract, 3) change the reward split via the `RewardsCoordinator` contract, 4) claim rewards posted by _AVSs_ from the `RewardsCoordinator` contract, and 5) change the allocation delay via the `AllocationManager` contract, this affects the delay until the allocation becomes slashable.

Author: deregistration is immediate, but still allocated funds are slashable for 14 days.

![Delegation](../diagrams/eigenlayer-delegation.png)

Withdrawal as a _Staker_ or _Restaker_ is only possible through the `DelegationManager` contract. When withdrawing the _Stakers_ and _Restakers_ receive the amount of deposit shares minus the amount of slashes suffered by the _AVS_ during their delegation. This amount of tokens are called withdrawable shares.

Withdrawal is a 2-step process which requires a `queueWithdrawal` and a `completeQueuedWithdrawal` call to the `DelegationManager` contract. The `queueWithdrawal` call is required to initiate the withdrawal process, and the `completeQueuedWithdrawal` call is required to complete the withdrawal process. The `completeQueuedWithdrawal` call is only possible after a cooldown period (14 days) is over. During this two step process, sufferred slashes by _AVS_ are taken into account to calculate the final amount of ETH to be sent to the _Restaker_.

> The `DelegationManager` contract also allows users to register as _Operators_ in the Eigenlayer system, which is a requirement for _Operators_ to register in _Operator Sets_ created by _AVSs_. _Operators_ can configure a _Delegation Approver_, that requires _Stakers_ to first get an approval before they may delegate to the _Operator_ through the `DelegationManager` contract. This way, _Operators_ can prevent _Stakers_ from delegating their funds to them and keep a privileged set of _Stakers_. If the _Delegation Approver_ is set to the 0-address, any _Staker_ can delegate their stake to the respective _Operator_.

Author: [3804415](https://github.com/deficollective/defiscan/pull/188/commits/380441589337c193ef04b600f483ecebb3dd4c5b)

separated one paragraph into two, probably simpler now

### AllocationManager

Through the `AllocationManager` the `AVSs` can manage _Operator Sets_ (creation and changes) and AVS configurations as metadata. The `AllocationManager` allows _Operators_ to allocate and deallocate from _Operator Sets_ and _AVSs_. _Operator Sets_ are isolated pools of staked capital that serve as economic security. _Operator Sets_ allow _AVSs_ to differentiate different _Operators_ with regarding to their business requirements and type of stake. Also _AVSs_ can slash certain strategies within an _Operator Set_ and issue rewards in a granular way for each supplied strategy within the _Operator Set_. Slashing is done when the _Operators_ break a certain committments to the _AVS_. _AVSs_ can slash _Operators_ up to the full stake inside the _Operator Sets_.

> _Operators_ can change their allocation delay which is the duration until which the allocated stake becomes slashable. A change of allocation delay registered by _Operator_ takes 17.5 days (ALLOCATION_CONFIGURATION_DELAY) until it is effective.

Author: because it's their terms when entering new AVSs, AVSs can upon registering by Operators reject them if they disagree with this delay. It does not affect, if they are registered/allocated already.

The _AVSs_ can setup an `AVSRegistrar` contract, which is called by the `AllocationManager` when _Operators_ register to and deregister from _Operator Sets_. If the _AVS_ does not have an `AVSRegistrar` set, the `AllocationManager` will forward the call to the registered _AVS_ contract. If the `AVSRegistrar` or the registered _AVS_ contract reverts the transaction, the `AllocationManager` will revert the transaction as well, and _Operators_ will not be able to register to or deregister from _Operator Sets_. This can create unfair scenarios, where _AVSs_ allow opt-in, but disallow opt-out (see [dependencies](#dependencies) for risks steming from _AVSs_).

The contract's _Upgradeability_ through the `ProxyAdmin` contract introduces an _Upgradeability_ risk, as the upgrade could potentially modify slashing logic and the slashing accounting and thus lead to _loss of funds_ of _(Re)Stakers_ if the upgrade is malicious or faulty, as the `DelegationManager` contract is informed on suffered slashes by the `AllocationManager` contract.

> Registration and deregistration of _Operators_ from _Operator Sets_, modification of allocations to _Operator Sets_ by _Operators_, slashing by _AVSs_ can be paused by [Pauser Multisig (1/6)](#security-council), [Executor Multisig (1/2)](#security-council) and [Operations Multisig (3/6)](#security-council), only the [Executor Multisig (1/2)](#security-council) can resume the mentioned actions. A pause of deregistration and modification of allocations but not slashing could lead to an unfair situation for _Operators_ and _Stakers_.

Author: i simplified, and left only the part that are different, ie functions + impact, who is mentioned earlier

> Registration and deregistration of _Operators_ from _Operator Sets_, modification of allocations to _Operator Sets_ by _Operators_, slashing by _AVSs_ can be paused by [Pauser Multisig (1/6)](#security-council), [Executor Multisig (1/2)](#security-council) and [Operations Multisig (3/6)](#security-council), only the [Executor Multisig (1/2)](#security-council) can resume the mentioned actions. A pause of deregistration and modification of allocations but not slashing could lead to an unfair situation for _Operators_ and _Stakers_.

Author: [757a4cf](https://github.com/deficollective/defiscan/pull/188/commits/757a4cf898b8f363d63895f014de5d571fbb5dd0)
